Switch BlockMap to use an unordered_set under the hood

JeremyRubin opened the pull request:
Currently we use an `unordered_map<uint256, CBlockIndex*>` for our BlockMap. This is a bit of a peculiar choice because we then store a pointer to the uint256 from the pair inside of the CBlockIndex. It would be conceptually simpler if we used a `unordered_map<CBlockIndex>` and modified `CBlockIndex` to directly own the uint256 for the phashBlock.

That's what this patch attempts doing. The only additional complexity is where we do want to query the map by key, the key equivalence stuff is only in C++20, so we add a helper that lets us construct a mock element easily for using with find.

The result is nicer since we get rid of an additional indirection for accessing the phashblock and we get rid of a lot of std::pair de-structuring. We also save a bit of memory per index (I haven't computed it precisely, but my guess is we save something like 24 bytes total per index by eliminating 1 pointer, which saves 8 bytes, and then could save us something like 16 bytes of padding). We also no longer can be in an inconsistent state where phashblock does not point to the entries own hash (although modifying phashblock once it is inserted into the map would be invalid -- it must be removed, modified, and reinserted).

Future work can likely further improve on this by making CBlockIndex owned* by the unordered_set directly, which will eliminate even more indirection/pointer chasing and make a lot of the code using BlockMap simpler (I don't think we ever rely on CBlockIndex being non-owned anywhere, nor should we have reason to in the future).


* see https://en.cppreference.com/w/cpp/container/unordered_set, it will still be safe to use pointers for e.g., pprev.
```
     References and pointers to data stored in the container are only invalidated by erasing that element, even when the corresponding iterator is invalidated. 
```

JeremyRubin committed:
Refactor BlockMap to use an unordered_set instead of an unordered_map

JeremyRubin commented on src/validation.cpp:
```diff
@@ -1875,8 +1876,8 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consens
     if (consensusparams.BIP16Exception.IsNull() || // no bip16 exception on this chain
-        pindex->phashBlock == nullptr || // this is a new candidate block, eg from TestBlockValidity()
-        *pindex->phashBlock != consensusparams.BIP16Exception) // this block isn't the historical exception
+        pindex->m_hash_block.IsNull() || // this is a new candidate block, eg from TestBlockValidity()
```
is this kosher 100% of the time as a replacement for nullptr?

MarcoFalke replied:
This should fix itself in the next rebase?

promag commented on src/chain.h:
```diff
@@ -232,7 +232,7 @@ class CBlockIndex
 
     uint256 GetBlockHash() const
```
nit, could change to `const uint256&`? Or drop it?

JeremyRubin replied:
done!

promag reviewed (commented):
Concept ACK!

>  This is a bit of a peculiar choice because we then store a pointer to the uint256 from the pair inside of the CBlockIndex

Indeed.

JeremyRubin commented:
Thanks for the review! Converted GetBlockHash to be a const ref, good suggestion!

I've removed the WIP status here as I eliminated the earlier bug.

promag commented on src/test/fuzz/chain.cpp:
```diff
@@ -20,7 +20,7 @@ void test_one_input(const std::vector<uint8_t>& buffer)
 
     const uint256 zero{};
-    disk_block_index->phashBlock = &zero;
+    disk_block_index->m_hash_block = zero;
```
Could remove? Already null right?

JeremyRubin replied:
done

promag commented on src/test/fuzz/chain.cpp:
```diff
@@ -59,7 +59,7 @@ void test_one_input(const std::vector<uint8_t>& buffer)
 
     CBlockIndex block_index{block_header};
-    block_index.phashBlock = &zero;
+    block_index.m_hash_block = zero;
```
Could be removed.

JeremyRubin replied:
done

JeremyRubin commented on src/chain.h:
```diff
@@ -230,9 +230,9 @@ class CBlockIndex
     }
 
-    uint256 GetBlockHash() const
+    const uint256& GetBlockHash() const
```
Better for follow up work; this is called in a huge number of places throughout the codebase.

JeremyRubin commented on src/bench/rpc_blockchain.cpp:
```diff
@@ -22,7 +22,7 @@ static void BlockToJsonVerbose(benchmark::Bench& bench)
     CBlockIndex blockindex;
     const uint256 blockHash = block.GetHash();
-    blockindex.phashBlock = &blockHash;
+    blockindex.m_hash_block = blockHash;
```
done

JeremyRubin commented on src/chain.h:
```diff
@@ -138,7 +138,7 @@ class CBlockIndex
 {
 public:
     //! pointer to the hash of the block, if any. Memory is owned by this CBlockIndex
```
done

JeremyRubin commented:
if there is a concept ACK from someone who would merge it on this approach I will rebase it, but since it touches a lot of places I don't have time to rebase it continually.

MarcoFalke commented:
Yeah, it is non-trivial to review so hard to tell.

I wonder if this tiny layers can be peeled off of this or if this needs to be one chunk. Maybe the `phashBlock` change can be split out?

JeremyRubin commented:
hmmm...

it's possible? but why?

I see two split-up paths:

### Make phashblock m_hash_block and keep the map structure first

This uses 24 bytes more memory at first, but doesn't change the lookup data structure.

Next, we could change the data structure to an indirect_map type and change the key to a pointer.

Then, we could change to a set.

### Make the data structure a set first and keep phashblock

This changes the data structure, and changes all the lookup sites.

No extra storage, no savings.

Uses of phashblock remain consistent, although it's just a self-pointer.

phashblock could become an inline method too, so you don't change the memory usage, but do need to add ().

Eventually switch to using m_hash_block.

MarcoFalke commented:
Is this still relevant after #24050?

JeremyRubin commented:
yes, it's completely orthogonal to https://github.com/bitcoin/bitcoin/pull/24050

ryanofsky commented on src/validation.h:
```diff
@@ -101,7 +101,21 @@ struct BlockHasher
+    size_t operator()(CBlockIndex* const& ptr) const { return ReadLE64(ptr->GetBlockHash().begin()); }
+    // Helper for querying by hash
+    // e.g., map.find(map.hash_function()(h))
+    CBlockIndex* operator()(const uint256& hash) {
```
In commit "Refactor BlockMap to use an unordered_set instead of an unordered_map" (fdc2f15c6b821f8ac0936cd35272a37305801723)

I was confused by the `idx.hash_function()(hash)` stuff initially, and think it could be clearer if you gave this method a different name like `FakeBlockIndex` instead of `operator()`. `idx.hash_function().FakeBlockIndex(hash)` would give a clue that the method is returning a fake block index pointer, not taking the hash of a hash, and also avoid overloading the same operator in two ways that do different things conceptually.

Also I guess you are using a shared mock CBlockIndex instance rather than creating temporary CBlockIndex objects for performance reasons, but it it would be interesting to know if taking a more straightforward with temporary mock objects would actually hurt performance:

```c++
struct FakeBlockIndex : public CBlockIndex
{
    FakeBlockIndex(const uint256& hash) { m_hash_block = hash; }
    FakeBlockIndex* pointer() { return this; }
};
```

Since it would allow simplifying:

```diff
-        BlockMap::const_iterator  it = m_blockman.m_block_index.find(m_blockman.m_block_index.hash_function()(hashAssumeValid));
+        BlockMap::const_iterator  it = m_blockman.m_block_index.find(FakeBlockIndex(hashAssumeValid).pointer());
```

and:

```diff
 struct BlockHasher
 {
     // this used to call `GetCheapHash()` in uint256, which was later moved; the
     // cheap hash function simply calls ReadLE64() however, so the end result is
     // identical
-    mutable CBlockIndex mock;
-    BlockHasher() : mock() {};
     size_t operator()(CBlockIndex* const& ptr) const { return ReadLE64(ptr->GetBlockHash().begin()); }
-    // Helper for querying by hash
-    // e.g., map.find(map.hash_function()(h))
-    CBlockIndex* operator()(const uint256& hash) {
-        mock.m_hash_block = hash;
-        return &mock;
-    }
 };
```

JeremyRubin replied:
I think it would since CBlockIndex is a very big struct?

but maybe we can make CBlockIndex have an abstract base class with just m_block_hash and then inherit from that for both? but then you have overhead *everywhere*...

ryanofsky replied:
> I think it would since CBlockIndex is a very big struct?
> 
> but maybe we can make CBlockIndex have an abstract base class with just m_block_hash and then inherit from that for both? but then you have overhead _everywhere_...

I would probably not do the abstract base thing, since the reason for my suggestion was to simplify code, and trying to overload find that way would seem to add more complexity. Current approach here does seems fine. If dropping the hasher method would be bad for performance, I'd just rename it from `operator()` to `FakeBlockIndex` or something for clarity.

JeremyRubin replied:
i probably won't have time to work on this until mid april FYI, but i can pick it up sometime then.

if you'd like to pick it off and rebase it and push it through i would review it + ack.

ryanofsky replied:
Good to know. I'm working on some other CBlockIndex changes so might want to do this. No hurry though!

ryanofsky reviewed (commented):
Concept ACK. I started making the same change after reviewing #24050 and #24199. As noted in the PR description, C++20 will allow a simpler implementation of this because it adds `unordered_set` [`find`](https://www.cppstories.com/2021/heterogeneous-access-cpp20/) and `count` method overloads that can accept block hashes, instead of CBlockIndex pointers (heterogenous lookups, https://www.cppstories.com/2021/heterogeneous-access-cpp20/). But since the find method is mostly wrapped anyway and not called directly, this is not a big deal.

Rspigler commented:
Mark up for grabs?
